polkadot/node/core/prospective-parachains/src/lib.rs

```diff
@@ -280,119 +306,142 @@ async fn handle_active_leaves_update<Context>(
 				compact_pending.push(c.compact);
 			}
 
-			let scope = FragmentChainScope::with_ancestors(
-				para,
-				block_info.clone(),
+			let scope = match FragmentChainScope::with_ancestors(
+				block_info.clone().into(),
 				constraints,
 				compact_pending,
 				max_candidate_depth,
-				ancestry.iter().cloned(),
-			)
-			.expect("ancestors are provided in reverse order and correctly; qed");
+				ancestry
+					.iter()
+					.map(|a| RelayChainBlockInfo::from(a.clone()))
+					.collect::<Vec<_>>(),
+			) {
+				Ok(scope) => scope,
+				Err(unexpected_ancestors) => {
+					gum::warn!(
+						target: LOG_TARGET,
+						para_id = ?para,
+						max_candidate_depth,
+						?ancestry,
+						leaf = ?hash,
+						"Relay chain ancestors have wrong order: {:?}",
+						unexpected_ancestors
+					);
+					continue
+				},
+			};
+
+			// Get the candidate storage of the parent leaf, if present.
+			let prev_fragment_chain = prev_fragment_chains.and_then(|chains| chains.get(&para));
+
+			if let Some(prev_fragment_chain) = prev_fragment_chain {
+				// Add old candidates to the new storage only after we added the pending
+				// availability candidates. The pending candidates have higher priority and can
+				// conflict with the old candidates.
+				for candidate in prev_fragment_chain.as_candidate_storage().into_candidates() {
+					// If they used to be pending availability, don't add them. This is fine
+					// because:
+					// - if they still are pending availability, they have already been added to the
+					//   new storage.
+					// - if they were included, no point in keeping them.
+					if prev_fragment_chain
+						.scope()
+						.get_pending_availability(&candidate.hash())
+						.is_none()
+					{
+						// We need to swallow any potential errors here, as they can happen under
+						// normal operation, with candidates becoming out of scope for example.
+						let _ = new_storage.add_candidate_entry(candidate);
```

Comment:
This function does not have access to scope and does not check it at all. The only thing it checks whether the candidate is present already.

Reply:
correct, the comment is misleading